src/programs/kmerprob.1
.TH KMERPROB 1 "July 2013" BIOBAMBAM
.SH NAME
kmerprob - print probability of a query sequence randomly appearing in a reference sequence
.SH SYNOPSIS
.PP
.B kmerprob
[options]
.SH DESCRIPTION
Let L be the length of a reference sequence. Further let Q be a query
sequence of length K such that Q contains Kx bases of type x for x=A,C,G,T
where KA+KC+KG+KT=K. Let px for x=A,C,G,T be the relative frequences of the
bases A,C,G,T in the reference sequence such that pA+pC+pG+pT=1. Then
kmerprob given the parameters L,KA,KC,KG,KT,pA,pC,pG,PT and n prints an
approximation of the probability that Q appears in such a random reference
sequence exactly n times.
.PP
The following key=value pairs can be given:
.PP
.B L: 
lenght of the reference sequence
.PP
.B KA:
number of A bases in the query sequence
.PP
.B KC:
number of C bases in the query sequence
.PP
.B KG:
number of G bases in the query sequence
.PP
.B KT:
number of T bases in the query sequence
.PP
.B pA;
relative frequency of the base A (i.e. number of times A appears in the reference divided by L)
.PP
.B pC;
relative frequency of the base C (i.e. number of times C appears in the reference divided by L)
.PP
.B pG;
relative frequency of the base G (i.e. number of times G appears in the reference divided by L)
.PP
.B pT;
relative frequency of the base T (i.e. number of times T appears in the reference divided by L)
.PP
.B n;
number of times query appears
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2013 German Tischler, \(co 2011-2013 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100

biobambam2 (2.0.22) unstable; urgency=medium

  * implicitly concatenate multiple input files as input for bamsort 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 29 Oct 2015 09:59:48 +0100

biobambam2 (2.0.21) unstable; urgency=medium

  * fix FF and RR type alignments in streaming duplicate marking
  * add MC tag (mate cigar string) when duplicate marking support is enabled in bamsort

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 26 Oct 2015 17:12:04 +0100

biobambam2 (2.0.20) unstable; urgency=medium

  * fix marking of FF and RR type duplicate records in BAM files 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 23 Oct 2015 14:55:46 +0200

biobambam2 (2.0.19) unstable; urgency=medium

  * include updates by Andrew Whitwham for improving the cigar handling of ClipAdapters

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 22 Sep 2015 16:51:25 +0200

biobambam2 (2.0.18) unstable; urgency=medium

  * update to libmaus2 2.0.68 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 14 Sep 2015 14:03:00 +0200

biobambam2 (2.0.17) unstable; urgency=medium

  * allow building without requirement for SSE4 (change in libmaus2) 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 03 Sep 2015 10:39:07 +0200

biobambam2 (2.0.16) unstable; urgency=medium

  * merge Andrew Whitwham's patch for removing adapter clipping tags from secondary and supplementary alignments 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 24 Aug 2015 13:15:02 +0200

biobambam2 (2.0.15) unstable; urgency=medium

  * update to libmaus2 2.0.45 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 21 Aug 2015 10:13:18 +0200

biobambam2 (2.0.14) unstable; urgency=medium

  * update to libmaus2 2.0.42 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 09 Aug 2015 13:36:32 +0200

biobambam2 (2.0.13) unstable; urgency=medium

  * allow in memory tmp files in bammarkduplicates, bammarkduplicates2 and bamtofastq 

 -- German Tischler <gt1@sanger.ac.uk>  Sat, 08 Aug 2015 15:03:19 +0200

biobambam2 (2.0.12) unstable; urgency=medium

  * update to libmaus2 2.0.40

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 07 Aug 2015 08:36:53 +0200

biobambam2 (2.0.11) unstable; urgency=medium

  * update to libmaus2 2.0.39 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 06 Aug 2015 18:23:03 +0200

biobambam2 (2.0.10) unstable; urgency=medium

  * include Andrew Whitwham's (streaming) duplicate marking tag changes

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 22 Jun 2015 11:22:34 +0200

biobambam2 (2.0.9) unstable; urgency=medium

  * add daligner las to SAM/BAM/CRAM conversion tool 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 19 Jun 2015 15:45:53 +0200

biobambam2 (2.0.8) unstable; urgency=low

  * fix for random failures in bamsormadup due to missing/incomplete initialisation 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 23:49:56 +0100

biobambam2 (2.0.7) unstable; urgency=low

  * update libmaus version to 2.0.6 (fix for bam indexing) 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 11:43:45 +0100

biobambam2 (2.0.6) unstable; urgency=low

  * update libmaus version to 2.0.5 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 24 Apr 2015 09:31:27 +0100

biobambam2 (2.0.5) unstable; urgency=low

  * fix/clarify manual page for bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 15:33:14 +0100

biobambam2 (2.0.4) unstable; urgency=low

  * add option to store temp data in RAM instead of on disc in bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 13:53:51 +0100

biobambam2 (2.0.3) unstable; urgency=medium

  * update to libmaus 2.0.3 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 12:48:32 +0100

biobambam2 (2.0.2) unstable; urgency=medium

  * update to libmaus 2.0.2 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 10:36:05 +0100

biobambam2 (2.0.1) unstable; urgency=medium

  * no change version bump to check version bumping script 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 00:15:29 +0100

biobambam2 (2.0.0) unstable; urgency=low

  * import of biobambam code

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 18 Apr 2015 12:00:00 +0100


src/programs/kmerprob.cpp
/**
    bambam
    Copyright (C) 2009-2013 German Tischler
    Copyright (C) 2011-2013 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include "config.h"

#include <iostream>
#include <queue>

#include <libmaus2/bambam/BamAlignment.hpp>
#include <libmaus2/bambam/BamDecoder.hpp>
#include <libmaus2/bambam/BamWriter.hpp>
#include <libmaus2/bambam/ProgramHeaderLineSet.hpp>

#include <libmaus2/util/ArgInfo.hpp>

#include <biobambam2/Licensing.hpp>
#include <biobambam2/KmerPoisson.hpp>

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);
		
		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if ( 
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if ( 
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;
				
				std::vector< std::pair<std::string,std::string> > V;
			
				V.push_back ( std::pair<std::string,std::string> ( "L", "length of reference sequence" ) );
				V.push_back ( std::pair<std::string,std::string> ( "n", "number of occurences" ) );
				V.push_back ( std::pair<std::string,std::string> ( "KA", "number of A bases in the query sequence" ) );
				V.push_back ( std::pair<std::string,std::string> ( "KC", "number of C bases in the query sequence" ) );
				V.push_back ( std::pair<std::string,std::string> ( "KG", "number of G bases in the query sequence" ) );
				V.push_back ( std::pair<std::string,std::string> ( "KT", "number of T bases in the query sequence" ) );
				V.push_back ( std::pair<std::string,std::string> ( "pA", "relative frequency of A base in reference" ) );
				V.push_back ( std::pair<std::string,std::string> ( "pC", "relative frequency of C base in reference" ) );
				V.push_back ( std::pair<std::string,std::string> ( "pG", "relative frequency of G base in reference" ) );
				V.push_back ( std::pair<std::string,std::string> ( "pT", "relative frequency of T base in reference" ) );
				
				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
								
				return EXIT_SUCCESS;
			}
			
		if ( !arginfo.hasArg("L") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need length of reference sequence (set key L)" << std::endl;
			se.finish();
			throw se;
		}
		if ( !arginfo.hasArg("KA") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need number of A bases in query (set key KA)" << std::endl;
			se.finish();
			throw se;
		}
		if ( !arginfo.hasArg("KC") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need number of C bases in query (set key KC)" << std::endl;
			se.finish();
			throw se;
		}
		if ( !arginfo.hasArg("KG") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need number of G bases in query (set key KG)" << std::endl;
			se.finish();
			throw se;
		}
		if ( !arginfo.hasArg("KT") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need number of T bases in query (set key KT)" << std::endl;
			se.finish();
			throw se;
		}
		if ( !arginfo.hasArg("n") )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "Need number of occurences (set key n)" << std::endl;
			se.finish();
			throw se;
		}
		
		uint64_t const L = arginfo.getValue<uint64_t>("L",0);
		uint64_t const KA = arginfo.getValue<uint64_t>("KA",0);
		uint64_t const KC = arginfo.getValue<uint64_t>("KC",0);
		uint64_t const KG = arginfo.getValue<uint64_t>("KG",0);
		uint64_t const KT = arginfo.getValue<uint64_t>("KT",0);
		uint64_t const n = arginfo.getValue<uint64_t>("n",0);
		double const pA = arginfo.getValue<double>("pA",0.25);
		double const pC = arginfo.getValue<double>("pC",0.25);
		double const pG = arginfo.getValue<double>("pG",0.25);
		double const pT = arginfo.getValue<double>("pT",0.25);

		double const p = pA + pC + pG + pT;
		double const eps = 1e-4;
		
		if ( p < 1.0 - eps || p > 1.0 + eps )
		{
			libmaus2::exception::LibMausException se;
			se.getStream() << "pA + pC + pG + pT = " << p << " which is not within eps=" << eps << " of 1" << std::endl;
			se.finish();
			throw se;			
		}

		std::cout << kmerPoisson(L,KA,KC,KG,KT,n,pA,pC,pG,pT) << std::endl;
			
		return EXIT_SUCCESS;
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}

